File: lib/agent_blueprint_protocol/extension.ex
defmodule AgentBlueprintProtocol.Extension do
  @moduledoc """
  The positional extension envelope shared by BOTH artifact tables:
  `{"critical": {namespace → payload}, "optional": {namespace → payload}}`,
  digest-covered, with namespace form and total cardinality. Extracted from
  Blueprint so the Deployment table carries the same judgment as
  DATA instead of a copy of it — one envelope, one definition, two tables.

  Negotiation keeps its own region walk (its judgments — lifecycle states,
  schema pins — are different from the table's shape checks); this module is
  the TABLE-check and scan-walk surface only.
  Extension envelopes are inert data — an extension never authorizes an operation.
  Extension envelopes are inert data — an extension never authorizes an operation.
  """

  alias AgentBlueprintProtocol.Json

  @identifier_bytes 512
  @ceiling_extensions 32

  @namespace ~r/\A[a-z0-9][a-z0-9.-]*\/[a-z0-9][a-z0-9.-]*\z/

  @doc """
  The envelope shape check a table carries: exactly critical + optional
  regions, namespace form, total cardinality, no namespace in both.
  """
  @spec envelope_ok?(Json.value()) :: :ok | {:error, term()}
  def envelope_ok?({:object, members}) do
    with :ok <- shape(members),
         {:ok, namespaces} <- namespaces(members),
         :ok <- cardinality(namespaces) do
      no_duplicates(namespaces)
    end
  end

  def envelope_ok?(_not_an_object), do: {:error, :invalid_type}

  @doc """
  Every `{namespace, body}` pair in the envelope, both regions, in document
  order — the scan's walk over the open region.
  """
  @spec bodies(list() | map()) :: [{binary(), Json.value()}]
  def bodies(members) do
    for {:object, regions} <- List.wrap(member_value(members, "extensions")),
        {_criticality, {:object, namespaces}} <- regions,
        {ns, body} <- namespaces,
        do: {ns, body}
  end

  @doc """
  The critical region's namespace set for THIS value — the authored-channel
  tie (a validated-critical list must sit inside this set).
  """
  @spec critical_namespaces(list() | map()) :: MapSet.t()
  def critical_namespaces(members) do
    for {:object, regions} <- List.wrap(member_value(members, "extensions")),
        {"critical", {:object, namespaces}} <- regions,
        {ns, _body} <- namespaces,
        do: ns,
        into: MapSet.new()
  end

  # ---- internals ------------------------------------------------------------------

  defp shape(members) do
    names = Enum.map(members, fn {name, _} -> name end)

    if Enum.sort(names) != ["critical", "optional"] do
      # A single missing half is a missing member; anything else unknown.
      if names == ["critical"] or names == ["optional"],
        do: {:error, :missing_required_field},
        else: {:error, :unknown_member}
    else
      :ok
    end
  end

  defp namespaces(members) do
    entries =
      Enum.flat_map(members, fn
        {_region, {:object, ns_members}} when is_list(ns_members) ->
          Enum.map(ns_members, fn {ns, body} -> namespace_entry(ns, body) end)

        {_region, _not_an_object} ->
          [{:error, :invalid_type}]
      end)

    first_error = Enum.find(entries, &match?({:error, _}, &1))

    if first_error,
      do: first_error,
      else: {:ok, Enum.map(entries, fn {:ok, ns, _body} -> ns end)}
  end

  defp namespace_entry(ns, body) do
    if namespace_form?(ns), do: {:ok, ns, body}, else: {:error, :extension_namespace_invalid}
  end

  defp namespace_form?(ns) do
    byte_size(ns) <= @identifier_bytes and Regex.match?(@namespace, ns)
  end

  defp cardinality(namespaces) do
    if length(namespaces) <= @ceiling_extensions,
      do: :ok,
      else: {:error, :invalid_cardinality}
  end

  defp no_duplicates(namespaces) do
    if length(namespaces) == length(Enum.uniq(namespaces)),
      do: :ok,
      else: {:error, :extension_duplicate}
  end

  # Root member lists only: both artifact call sites run against the ordered
  # member list (the scan walk and the authored-channel tie), never the
  # engine's hook-shaped map.
  defp member_value(members, name) when is_list(members) do
    case List.keyfind(members, name, 0) do
      {^name, value} -> value
      nil -> nil
    end
  end
end

File: conformance/verifier/extension.ts
// The positional extension envelope — the mirror of AgentBlueprintProtocol.Extension:
// {"critical": {ns → body}, "optional": {ns → body}}, shared by both artifact
// tables as a table check and by the scans as the open-region walk.

import type { Value } from "./value.ts";

const IDENTIFIER_BYTES = 512;
const CEILING_EXTENSIONS = 32;

const NAMESPACE = /^[a-z0-9][a-z0-9.-]*\/[a-z0-9][a-z0-9.-]*$/;

type VResult = { ok: true } | { ok: false; e: string };

// The envelope shape check a table carries: exactly critical + optional
// regions, namespace form, total cardinality, no namespace in both.
export function envelopeOk(value: Value): VResult {
  if (value.t !== "obj") return { ok: false, e: "invalid_type" };
  const members = value.v;

  const names = members.map(([name]) => name).sort();
  if (names.length !== 2 || names[0] !== "critical" || names[1] !== "optional") {
    // A single missing half is a missing member; anything else unknown.
    const raw = members.map(([name]) => name);
    if (raw.length === 1 && (raw[0] === "critical" || raw[0] === "optional")) {
      return { ok: false, e: "missing_required_field" };
    }
    return { ok: false, e: "unknown_member" };
  }

  const namespaces: string[] = [];
  for (const [, regionValue] of members) {
    if (regionValue.t !== "obj") return { ok: false, e: "invalid_type" };
    for (const [ns] of regionValue.v) {
      if (!namespaceForm(ns)) return { ok: false, e: "extension_namespace_invalid" };
      namespaces.push(ns);
    }
  }

  if (namespaces.length > CEILING_EXTENSIONS) return { ok: false, e: "invalid_cardinality" };

  if (new Set(namespaces).size !== namespaces.length) return { ok: false, e: "extension_duplicate" };

  return { ok: true };
}

// Every {namespace, body} pair in the envelope, both regions, in document
// order — the scan's walk over the open region. `members` is the ROOT member
// list (the "extensions" member is located by name).
export function bodies(members: [string, Value][]): [string, Value][] {
  const out: [string, Value][] = [];
  const extensions = memberValue(members, "extensions");
  if (extensions === null || extensions.t !== "obj") return out;
  for (const [, regionValue] of extensions.v) {
    if (regionValue.t !== "obj") continue;
    for (const [ns, body] of regionValue.v) out.push([ns, body]);
  }
  return out;
}

// The critical region's namespace set for THIS value — the authored-channel
// tie (a validated-critical list must sit inside this set).
export function criticalNamespaces(members: [string, Value][]): Set<string> {
  const out = new Set<string>();
  const extensions = memberValue(members, "extensions");
  if (extensions === null || extensions.t !== "obj") return out;
  for (const [region, regionValue] of extensions.v) {
    if (region !== "critical" || regionValue.t !== "obj") continue;
    for (const [ns] of regionValue.v) out.add(ns);
  }
  return out;
}

function namespaceForm(ns: string): boolean {
  return Buffer.byteLength(ns, "utf8") <= IDENTIFIER_BYTES && NAMESPACE.test(ns);
}

function memberValue(members: [string, Value][], name: string): Value | null {
  const found = members.find(([key]) => key === name);
  return found ? found[1] : null;
}
